lib/at_parser/at_parser.c

```diff
@@ -0,0 +1,411 @@
+
+#include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <errno.h>
+#include <zephyr/sys/util.h>
+#include <modem/at_parser.h>
+
+#include "at_token.h"
+#include "at_match.h"
+
+enum at_num_type {
+	AT_NUM_INT16,
+	AT_NUM_UINT16,
+	AT_NUM_INT32,
+	AT_NUM_UINT32,
+	AT_NUM_INT64,
+};
+
+static bool is_crlf(const char *str)
+{
+	return strlen(str) == 2 && str[0] == '\r' && str[1] == '\n';
+}
+
+static bool is_subparam(const struct at_token *token)
+{
+	switch (token->type) {
+	case AT_TOKEN_TYPE_INT:
+	case AT_TOKEN_TYPE_QUOTED_STRING:
+	case AT_TOKEN_TYPE_ARRAY:
+	case AT_TOKEN_TYPE_EMPTY:
+		return true;
+	default:
+		return false;
+	}
+}
+
+static bool is_resp(const char *str)
+{
+	static const char * const resp[] = {
+		"\r\nOK\r\n",
+		"\r\nERROR\r\n",
+		"\r\n+CME ERROR",
+		"\r\n+CMS ERROR"
+	};
+
+	for (size_t i = 0; i < ARRAY_SIZE(resp); i++) {
+		if (strncmp(str, resp[i], strlen(resp[i])) == 0) return true;
+	}
+
+	return false;
+}
+
+static bool is_index_ahead(struct at_parser *parser, size_t index)
+{
+	return index + 1 > parser->count;
+}
+
+static bool lookahead_crlf_or_more(const char *str)
+{
+	return strlen(str) >= 2 && str[0] == '\r' && str[1] == '\n';
+}
+
+static bool lookahead_crlf_and_more(const char *str)
+{
+	return strlen(str) > 2 && str[0] == '\r' && str[1] == '\n';
+}
+
+static void at_token_set_empty(struct at_token *token, const char *start)
+{
+	token->start = start;
+	token->len = 0;
+	token->type = AT_TOKEN_TYPE_EMPTY;
+	token->variant = AT_TOKEN_VARIANT_NO_COMMA;
+}
+
+int at_parser_init(struct at_parser *parser, const char *at)
+{
+	if (!parser || !at) {
+		return -EINVAL;
+	}
+
+	size_t len = strlen(at);
+
+	/* Check bounds. Minimum limit of 2 characters ("AT"). Maximum limit of 65535 characters. */
+	if (len < 2 || len > UINT16_MAX) {
+		return -EINVAL;
+	}
+
+	memset(parser, 0, sizeof(struct at_parser));
+
+	parser->cursor = at;
+	parser->is_init = true;
+
+	return 0;
+}
+
+static int at_parser_tok(struct at_parser *parser, struct at_token *token)
+{
+	const char *remainder = NULL;
+	size_t cursor_len = 0;
+	size_t remainder_len = 0;
+
+	cursor_len = strlen(parser->cursor);
+
+	/* The lexer cannot match empty strings, so intercept the special case where the empty
+	 * subparameter is the one after the previously parsed token.
+	 * This case is detected in the previous call to this function.
+	 */
+	if (parser->is_next_empty) {
+		at_token_set_empty(token, parser->cursor);
+
+		/* Set the remainder to the current cursor because an empty token has length
+		 * zero.
+		 */
+		remainder = parser->cursor;
+
+		/* Reset the flag. */
+		parser->is_next_empty = false;
+
+		goto matched;
+	}
+
+	/* Nothing left to tokenize. */
+	if (cursor_len == 0 || is_crlf(parser->cursor)) {
+		return -EIO;
+	}
+
+	/* Don't tokenize response. */
+	if (is_resp(parser->cursor)) {
+		return -EIO;
+	}
+
+	/* Check if there is a new line to parse. */
+	if (lookahead_crlf_and_more(parser->cursor)) {
+		return -EAGAIN;
```

Comment:
Why do we stop tokenizing on new lines? I think the user should be able to call this function in a loop without looking for special errors that mean "there is another line to parse". The parser should just start tokenizing the next line. If there is a use-case where the use need a function to parse one line at a time, I suggest to introduce a `_tok_line()` function.

Reply:
The parsing works line by line and is index based. Parsing without any bounds for lines would compromise this balance. A notification has 5 subparameters and outputs 2 lines of it. We ask for subparameter with index 8, what is it? What if the first line has 4 subparameters because it is not including the last optional? I appreciated the original design choice of returning `-EAGAIN` because it makes sense and informs the user, at runtime, that there is more to parse.

Reply:
ok